File: s5-problem-solving-patterns/freqency_counter/anagrams/solution2.js
const validAnagram = function (word1, word2) {
  if (word1.length !== word2.length) return false;

  const lookup = {};
  for (let l of word1) {
    lookup[l] ? lookup[l]++ : (lookup[l] = 1);
  }

  for (let l of word2) {
    if (!lookup[l]) return false;
    lookup[l]--; // Not when letter = 0; falsy value -> earlier if statement will fail
  }
  return true;
};

console.log(validAnagram("", "")); // true
console.log(validAnagram("aaz", "zza")); // false
console.log(validAnagram("rat", "car")); // false
console.log(validAnagram("anagram", "nagaram")); // true
console.log(validAnagram("awesome", "awesom")); // false
console.log(validAnagram("qwerty", "qeywrt")); // true

File: s5-problem-solving-patterns/freqency_counter/anagrams/solution2.py
def validAnagram(word1, word2):
    if len(word1) != len(word2):
        return False

    lookup = {}
    for l in word1:
        if l in lookup.keys():
            lookup[l] += 1
        else:
            lookup[l] = 1

    for l in word2:
        if l not in lookup.keys() or lookup[l] == 0:
            return False
        lookup[l] -= 1
    return True


print(validAnagram("", ""))  # true
print(validAnagram("aaz", "zza"))  # false
print(validAnagram("rat", "car"))  # false
print(validAnagram("anagram", "nagaram"))  # true
print(validAnagram("awesome", "awesom"))  # false
print(validAnagram("qwerty", "qeywrt"))  # true
